### ISSUE TITLE
[FEATURE] Get all metadata of AiService & Tool when the Spring Boot starts

### ISSUE THREAD

**Issue by catofdestruction** [langchain4j/langchain4j#2112]
First of all, I really like our framework!

Here are my specific needs:

I want to obtain the metadata of all @AiService annotations and their associated toolsets when the Spring Boot service starts. For example:

- Information about the chat method in AiService.
- Metadata of the tool list in AiService, such as List<ToolSpecification>.

The reason is that I might need to register this metadata to a management platform at service startup for reference and sharing.

Currently, I have already retrieved the methods related to AiService at startup, 
but I am unable to get the tool-related content through reflection. 
They should be stored in the AiServiceContext within each original AiService. 

Is there a good way to achieve this? Thank you very much.

**Comment by langchain4j** [comment]
Hi @catofdestruction, thank you for kind words!

Right now there is no way to do so, but you can take a look at these classes, maybe it will help somehow:
- https://github.com/langchain4j/langchain4j-spring/blob/main/langchain4j-spring-boot-starter/src/main/java/dev/langchain4j/service/spring/AiServicesAutoConfig.java
- https://github.com/langchain4j/langchain4j-spring/blob/main/langchain4j-spring-boot-starter/src/main/java/dev/langchain4j/service/spring/AiServiceFactory.java

Would you like to implement this feature?

BTW, there is a related issue: https://github.com/langchain4j/langchain4j/issues/924

**Comment by catofdestruction** [comment]
> Hi @catofdestruction, thank you for kind words!
> 
> Right now there is no way to do so, but you can take a look at these classes, maybe it will help somehow:
> 
> * https://github.com/langchain4j/langchain4j-spring/blob/main/langchain4j-spring-boot-starter/src/main/java/dev/langchain4j/service/spring/AiServicesAutoConfig.java
> * https://github.com/langchain4j/langchain4j-spring/blob/main/langchain4j-spring-boot-starter/src/main/java/dev/langchain4j/service/spring/AiServiceFactory.java
> 
> Would you like to implement this feature?
> 
> BTW, there is a related issue: #924

Thank you for your response. It would be an honor for me to have the opportunity to contribute code to the framework! I am also very eager to become a contributor to langchain4j.

Unfortunately, I currently lack experience in submitting code to open-source repositories on GitHub. Are there any detailed contribution steps I could refer to? I will follow the steps and try to contribute.

Here is the idea I have for implementing this feature:

1. Before the getObject method in the AiServiceFactory class returns, publish a Spring Event, for example, AiServiceContextReadyEvent.
2. The callback properties in AiServiceContextReadyEvent might include Class<?> aiServiceClass and List<ToolSpecification> toolSpecifications, or directly return AiServiceContext.
3. AiServiceFactory implements ApplicationEventPublisherAware to publish AiServiceContextReadyEvent, and users can listen to AiServiceContextReadyEvent in their own projects.

**Comment by langchain4j** [comment]
@catofdestruction that's great! You can find the contribution guidelines [here](https://github.com/langchain4j/langchain4j/blob/main/CONTRIBUTING.md).

Regarding the implementation, sounds good, although I have a few comments:
- I am not sure about exposing `AiServiceContext` directly, it is meant to be an internal implementation detail that I (as a maintainer) would like to have the flexibility to change any time. So perharps we could add all the required data to the `AiServiceReadyEvent`?
- I am not that familiar with Spring events, so it is hard to comment. Are there any other alternatives?

Thank you!

**Comment by catofdestruction** [comment]
> @catofdestruction that's great! You can find the contribution guidelines [here](https://github.com/langchain4j/langchain4j/blob/main/CONTRIBUTING.md).那太棒了！您可以[在此处](https://github.com/langchain4j/langchain4j/blob/main/CONTRIBUTING.md)找到贡献指南。
> 
> Regarding the implementation, sounds good, although I have a few comments:关于实施，听起来不错，但我有一些评论：
> 
> * I am not sure about exposing `AiServiceContext` directly, it is meant to be an internal implementation detail that I (as a maintainer) would like to have the flexibility to change any time. So perharps we could add all the required data to the `AiServiceReadyEvent`?我不确定是否直接公开`AiServiceContext` ，它是一个内部实现细节，我（作为维护者）希望能够随时灵活更改。那么我们可以将所有必需的数据添加到`AiServiceReadyEvent`中吗？
> * I am not that familiar with Spring events, so it is hard to comment. Are there any other alternatives?我对Spring事件不太熟悉，所以很难发表评论。还有其他选择吗？
> 
> Thank you! 谢谢你！

Thank you very much! Once I have the code ready, I will reply here as well.

**Comment by catofdestruction** [comment]
> @catofdestruction that's great! You can find the contribution guidelines [here](https://github.com/langchain4j/langchain4j/blob/main/CONTRIBUTING.md).
> 
> Regarding the implementation, sounds good, although I have a few comments:
> 
> * I am not sure about exposing `AiServiceContext` directly, it is meant to be an internal implementation detail that I (as a maintainer) would like to have the flexibility to change any time. So perharps we could add all the required data to the `AiServiceReadyEvent`?
> * I am not that familiar with Spring events, so it is hard to comment. Are there any other alternatives?
> 
> Thank you!

Hi, I followed the steps in the contribution guide and first forked a repository.

Currently, my solution is to publish a Spring Event after registering the AiService Bean in AiServicesAutoConfig. 
This event contains the AiService class and its corresponding tools description information.

Once a user implements the AiServiceRegisteredEventListener to listen for this event, they can receive the event during the Spring Boot startup phase and handle their business logic as needed.


For example, in the test class AiServiceWithToolsApplication, the onApplicationEvent method 
will receive the event at startup and print the following information:


```
@SpringBootApplication
class AiServiceWithToolsApplication implements AiServiceRegisteredEventListener {

    public static void main(String[] args) {
        SpringApplication.run(AiServiceWithToolsApplication.class, args);
    }

    @Override
    public void onApplicationEvent(AiServiceRegisteredEvent event) {
        Class<?> aiServiceClass = event.getAiServiceClass();
        List<ToolSpecification> toolSpecifications = event.getToolSpecifications();
        for (int i = 0; i < toolSpecifications.size(); i++) {
            System.out.printf("[%s]: [Tool-%s]: %s%n", aiServiceClass.getSimpleName(), i + 1, toolSpecifications.get(i));
        }
    }
}
```


```
[AiServiceWithTools]: [Tool-1]: ToolSpecification { name = "getCurrentDate", description = null, parameters = null, toolParameters = null }
[AiServiceWithTools]: [Tool-2]: ToolSpecification { name = "getCurrentTime", description = null, parameters = null, toolParameters = null }
```



You should be able to see my code changes through this link
https://github.com/langchain4j/langchain4j-spring/compare/main...catofdestruction:langchain4j-spring:aiservice_created_event

Have a nice weekend! :)

**Comment by langchain4j** [comment]
Hi @catofdestruction, great, looks good from the first glance! I will be able to look deeper next week. BTW do we really need `AiServiceRegisteredEventListener` or can user declare their own classes that `extends ApplicationListener<AiServiceRegisteredEvent>`?
Once you will be ready, please submit a PR.
Thank you, have a nice weekend as well!

**Comment by misselvexu** [comment]
> > @catofdestruction that's great! You can find the contribution guidelines [here](https://github.com/langchain4j/langchain4j/blob/main/CONTRIBUTING.md).
> > Regarding the implementation, sounds good, although I have a few comments:
> > 
> > * I am not sure about exposing `AiServiceContext` directly, it is meant to be an internal implementation detail that I (as a maintainer) would like to have the flexibility to change any time. So perharps we could add all the required data to the `AiServiceReadyEvent`?
> > * I am not that familiar with Spring events, so it is hard to comment. Are there any other alternatives?
> > 
> > Thank you!
> 
> Hi, I followed the steps in the contribution guide and first forked a repository.
> 
> Currently, my solution is to publish a Spring Event after registering the AiService Bean in AiServicesAutoConfig. This event contains the AiService class and its corresponding tools description information.
> 
> Once a user implements the AiServiceRegisteredEventListener to listen for this event, they can receive the event during the Spring Boot startup phase and handle their business logic as needed.
> 
> For example, in the test class AiServiceWithToolsApplication, the onApplicationEvent method will receive the event at startup and print the following information:
> 
> ```
> @SpringBootApplication
> class AiServiceWithToolsApplication implements AiServiceRegisteredEventListener {
> 
>     public static void main(String[] args) {
>         SpringApplication.run(AiServiceWithToolsApplication.class, args);
>     }
> 
>     @Override
>     public void onApplicationEvent(AiServiceRegisteredEvent event) {
>         Class<?> aiServiceClass = event.getAiServiceClass();
>         List<ToolSpecification> toolSpecifications = event.getToolSpecifications();
>         for (int i = 0; i < toolSpecifications.size(); i++) {
>             System.out.printf("[%s]: [Tool-%s]: %s%n", aiServiceClass.getSimpleName(), i + 1, toolSpecifications.get(i));
>         }
>     }
> }
> ```
> 
> ```
> [AiServiceWithTools]: [Tool-1]: ToolSpecification { name = "getCurrentDate", description = null, parameters = null, toolParameters = null }
> [AiServiceWithTools]: [Tool-2]: ToolSpecification { name = "getCurrentTime", description = null, parameters = null, toolParameters = null }
> ```
> 
> You should be able to see my code changes through this link [langchain4j/langchain4j-spring@main...catofdestruction:langchain4j-spring:aiservice_created_event](https://github.com/langchain4j/langchain4j-spring/compare/main...catofdestruction:langchain4j-spring:aiservice_created_event)
> 
> Have a nice weekend! :)

From your requirements, you want to get all the extension services related to `@AiService` to access to the external management platform, Spring event mechanism is able to get some meta information and data asynchronously in the process of spring service initialization, but I think that these data are not enough in terms of completeness and credibility, in addition, if we look at it from the point of view of the external management platform, there must be not only the need to view but also the need for some control class operations. In addition, if you look at it from the perspective of an external management platform, there must be not only the need to view ,  there must be some control operation needs .
I have two ideas :
1、whether you can extend the `BeanFactoryPostProcessor` to achieve their own specific needs . 
2、if you just need to read the class, can you use `spring-boot-actuator`'s `Endpoints` to satisfy your needs, this kind of `AiServiceContext` will not have an impact.

**Comment by catofdestruction** [comment]
> > > @catofdestruction that's great! You can find the contribution guidelines [here](https://github.com/langchain4j/langchain4j/blob/main/CONTRIBUTING.md).
> > > Regarding the implementation, sounds good, although I have a few comments:
> > > 
> > > * I am not sure about exposing `AiServiceContext` directly, it is meant to be an internal implementation detail that I (as a maintainer) would like to have the flexibility to change any time. So perharps we could add all the required data to the `AiServiceReadyEvent`?
> > > * I am not that familiar with Spring events, so it is hard to comment. Are there any other alternatives?
> > > 
> > > Thank you!
> > 
> > 
> > Hi, I followed the steps in the contribution guide and first forked a repository.
> > Currently, my solution is to publish a Spring Event after registering the AiService Bean in AiServicesAutoConfig. This event contains the AiService class and its corresponding tools description information.
> > Once a user implements the AiServiceRegisteredEventListener to listen for this event, they can receive the event during the Spring Boot startup phase and handle their business logic as needed.
> > For example, in the test class AiServiceWithToolsApplication, the onApplicationEvent method will receive the event at startup and print the following information:
> > ```
> > @SpringBootApplication
> > class AiServiceWithToolsApplication implements AiServiceRegisteredEventListener {
> > 
> >     public static void main(String[] args) {
> >         SpringApplication.run(AiServiceWithToolsApplication.class, args);
> >     }
> > 
> >     @Override
> >     public void onApplicationEvent(AiServiceRegisteredEvent event) {
> >         Class<?> aiServiceClass = event.getAiServiceClass();
> >         List<ToolSpecification> toolSpecifications = event.getToolSpecifications();
> >         for (int i = 0; i < toolSpecifications.size(); i++) {
> >             System.out.printf("[%s]: [Tool-%s]: %s%n", aiServiceClass.getSimpleName(), i + 1, toolSpecifications.get(i));
> >         }
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > [AiServiceWithTools]: [Tool-1]: ToolSpecification { name = "getCurrentDate", description = null, parameters = null, toolParameters = null }
> > [AiServiceWithTools]: [Tool-2]: ToolSpecification { name = "getCurrentTime", description = null, parameters = null, toolParameters = null }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You should be able to see my code changes through this link [langchain4j/langchain4j-spring@main...catofdestruction:langchain4j-spring:aiservice_created_event](https://github.com/langchain4j/langchain4j-spring/compare/main...catofdestruction:langchain4j-spring:aiservice_created_event)
> > Have a nice weekend! :)
> 
> From your requirements, you want to get all the extension services related to `@AiService` to access to the external management platform, Spring event mechanism is able to get some meta information and data asynchronously in the process of spring service initialization, but I think that these data are not enough in terms of completeness and credibility, in addition, if we look at it from the point of view of the external management platform, there must be not only the need to view but also the need for some control class operations. In addition, if you look at it from the perspective of an external management platform, there must be not only the need to view , there must be some control operation needs . I have two ideas : 1、whether you can extend the `BeanFactoryPostProcessor` to achieve their own specific needs . 2、if you just need to read the class, can you use `spring-boot-actuator`'s `Endpoints` to satisfy your needs, this kind of `AiServiceContext` will not have an impact.

Currently, based on the langchain4j framework, my requirement is just to obtain the metadata for AiService and Tools.

I believe that the external management platform may have more registration information for AiService and Tools, such as URI, status, and so on. Those pieces of information do not seem to belong in the framework itself but should be extended within the business application and associated with the information defined by annotations in the framework. For example, the extension based on BeanFactoryPostProcessor that you mentioned.

Additionally, using Spring Boot Actuator’s Endpoints to obtain information about AiService is a good suggestion, and I will look into it further. Thank you!

**Comment by catofdestruction** [comment]
> Hi @catofdestruction, great, looks good from the first glance! I will be able to look deeper next week. BTW do we really need `AiServiceRegisteredEventListener` or can user declare their own classes that `extends ApplicationListener<AiServiceRegisteredEvent>`? Once you will be ready, please submit a PR. Thank you, have a nice weekend as well!

Hi, I removed the AiServiceRegisteredEventListener and placed it into the test case. Additionally, I added a test method in the AiServicesAutoConfigIT class to verify the reception of the event, and it has passed the test.

The code can be viewed here:
https://github.com/langchain4j/langchain4j-spring/compare/main…catofdestruction:langchain4j-spring:aiservice_created_event

Should I submit a PR to the main branch, or to another branch? Thank you!

<img width="1145" alt="image" src="https://github.com/user-attachments/assets/fa386e1d-717d-48af-a3e0-d502595e0e89">
